### PR TITLE
Module aliasing: validate source files / only allow Swift files for modules aliased

### PR DESCRIPTION
**Pull request by elsh**
Validate source files for modules being aliased to only allow Swift sources
Resolves rdar://90661600

### CODE REVIEW

**Comment by elsh** [comment]
@swift-ci smoke test

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 1]
```diff
@@ -710,31 +728,32 @@ private class ModuleAliasTracker {
...
-    func addAliasesForTargets(_ targets: [String],
+    func addAliasesForTargets(_ targets: [Target],
                               product: String,
                               package: PackageIdentity) {
-        
         let aliases = aliasMap[package]?[product]
-        for targetName in targets {
-            if idTargetToAliases[package]?[targetName] == nil {
-                idTargetToAliases[package, default: [:]][targetName] = []
+        for t in targets {
```
Comment (tomerd) [inline Sources/PackageGraph/PackageGraph+Loading.swift:735]: we normally try to use clear variable names, so `for target in targets`

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 2]
```diff
@@ -710,31 +728,32 @@ private class ModuleAliasTracker {
...
-               originPackage: PackageIdentity) -> String? {
-        if let aliasDict = aliasMap[originPackage] {
-            let models = aliasDict.values.flatMap{$0}.filter { $0.name == target }
-            // this func only checks if there's any existing alias so
-            // just return the first alias value
-            return models.first?.alias
+    func validateSources(_ targets: [Target],
+                         product: String,
+                         package: PackageIdentity) throws {
+        for t in targets {
```
Comment (tomerd) [inline Sources/PackageGraph/PackageGraph+Loading.swift:749]: ditto

**Comment by elsh** [comment]
@swift-ci  smoke test

**Comment by tomerd** [comment]
@swift-ci smoke test

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 3]
```diff
@@ -271,7 +271,7 @@ private func createResolvedPackages(
     }
 
     // Gather and resolve module aliases specified for targets in all dependent packages
-    let packageAliases = resolveModuleAliases(packageBuilders: packageBuilders,
+    let packageAliases = try resolveModuleAliases(packageBuilders: packageBuilders,
```
Comment (tomerd) [inline Sources/PackageGraph/PackageGraph+Loading.swift:274]: nit: formatting

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 4]
```diff
@@ -710,31 +728,32 @@ private class ModuleAliasTracker {
...
-            // this func only checks if there's any existing alias so
-            // just return the first alias value
-            return models.first?.alias
+    func validateSources(_ targets: [Target],
+                         product: String,
+                         package: PackageIdentity) throws {
+        for target in targets {
+            if let aliases = idTargetToAliases[package]?[target.name], !aliases.isEmpty {
+                let hasNonSwiftFiles = target.sources.containsNonSwiftFiles
+                if hasNonSwiftFiles {
```
Comment (tomerd) [inline Sources/PackageGraph/PackageGraph+Loading.swift:752]: nit: variable (hasNonSwiftFiles) seems redundant

File: Sources/PackageGraph/PackageGraph.swift [hunk 5]
```diff
@@ -227,6 +233,8 @@ extension PackageGraphError: CustomStringConvertible {
             return "multiple products named '\(product)' in: '\(packages.joined(separator: "', '"))'"
         case .multipleModuleAliases(let target, let product, let package, let aliases):
             return "multiple aliases: ['\(aliases.joined(separator: "', '"))'] found for target '\(target)' in product '\(product)' from package '\(package)'"
+        case .invalidSourcesForModuleAliasing(let target, let product, let package):
+            return "invalid sources for module aliasing; to enable it, target '\(target)' in product '\(product)' from package '\(package)' should only contain Swift source files"
```
Comment (tomerd) [inline Sources/PackageGraph/PackageGraph.swift:237]: maybe this should say something like "module aliasing can only be used swift based targets"

File: Tests/BuildTests/BuildPlanTests.swift [hunk 6]
```diff
@@ -1978,11 +1979,230 @@ final class BuildPlanTests: XCTestCase {
...
+                                          dependencies: ["Logging",
+                                                         .product(name: "Utils",
+                                                                  package: "fooPkg",
+                                                                  moduleAliases: ["Logging": "FooLogging"])
+                                          ]),
+                        TargetDescription(name: "Logging", dependencies: []),
+                    ]),
+            ],
+            observabilityScope: observability.topScope
+        ))
```
Comment (tomerd) [inline Tests/BuildTests/BuildPlanTests.swift:2205]: nice test coverage

**Review by tomerd — APPROVED** [review]

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 7]
```diff
@@ -694,7 +704,15 @@ private class ModuleAliasTracker {
                   target: String,
                   product: String,
                   originPackage: PackageIdentity,
-                  consumingPackage: PackageIdentity) {
+                  consumingPackage: PackageIdentity) throws {
+        if let aliasDict = aliasMap[originPackage] {
+            let models = aliasDict.values.flatMap{$0}.filter { $0.name == target }
+            if !models.isEmpty {
+                // Error if there are multiple aliases specified for this product dependency
+                throw PackageGraphError.multipleModuleAliases(target: target, product: product, package: originPackage.description, aliases: models.map{$0.alias} + [alias])
```
Comment (abertelrud) [inline Sources/PackageGraph/PackageGraph+Loading.swift:712]: If I had more than one conflicting alias, I'd see an error about one at a time, right?  I think that's fine, just checking if I understand the flow.

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 8]
```diff
@@ -668,15 +663,30 @@ private func resolveModuleAliases(packageBuilders: [ResolvedPackageBuilder],
...
     }
 
     // Override module aliases upstream if needed
     aliasTracker.propagateAliases()
 
+    // Validate sources (Swift files only) for modules being aliased.
+    // Needs to be done after `propagateAliases` since aliases defined
+    // upstream can be overriden.
+    for packageBuilder in packageBuilders {
+        for produdct in packageBuilder.package.products {
```
Comment (abertelrud) [inline Sources/PackageGraph/PackageGraph+Loading.swift:681]:
```suggestion
        for product in packageBuilder.package.products {
```

File: Sources/PackageGraph/PackageGraph+Loading.swift [hunk 9]
```diff
@@ -668,15 +663,30 @@ private func resolveModuleAliases(packageBuilders: [ResolvedPackageBuilder],
     // Track targets that need module aliases for each package
     for packageBuilder in packageBuilders {
         for produdct in packageBuilder.package.products {
```
Comment (abertelrud) [inline Sources/PackageGraph/PackageGraph+Loading.swift:665]:
```suggestion
        for product in packageBuilder.package.products {
```

File: Sources/PackageGraph/PackageGraph+Loading.swift (hunk 7 above)
Comment (elsh) [reply Sources/PackageGraph/PackageGraph+Loading.swift:712]: this error will show all conflicting aliases defined for that target in the product

**Comment by elsh** [comment]
@swift-ci smoke test

**Review by neonichu — APPROVED** [review]

**Review by abertelrud — APPROVED** [review]

**Pull request merged**